### PR TITLE
rfc: User-defined SQL schemas

### PR DESCRIPTION
**Pull request by knz**
Informs #30916.
Release note: None

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
A very comprehensive RFC!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-:-LNu7nY0EiaJGrViBvIF:ba0qm5g)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNu6Vcg7NwJj6fp9z4y:b54spep) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
> ```Markdown
> 
> (I like the idea to avoid descriptors because that would ask complex
> questions about leases and the allocation of ranges.)
> ```

Can you elaborate on this concern? I was imagining that stored schemas would be implemented with some sort of `SchemaDescriptor` and would be present in the `system.namespace` table very much like you describe in this section. How would any of these alternatives affect the allocation of ranges (or range IDs which you allude to elsewhere)? Are the leases you're concerned about table leases? We don't currently have leases on databases (unless something changed that I'm not aware of).

`system.namespace` was initially designed to support arbitrary hierarchy similar to a filesystem, but we limited the initial lookup support to only 2 levels (database, table). I think you might be aware of this.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuCv8G5fyzb6hCPbdy:b-m58iqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you elaborate on this concern? I was imagining that stored schemas would be implemented with some sort of `SchemaDescriptor` and would be present in the `system.namespace` table very much like you describe in this section. How would any of these alternatives affect the allocation of ranges (or range IDs which you allude to elsewhere)? Are the leases you're concerned about table leases? We don't currently have leases on databases (unless something changed that I'm not aware of).

`system.namespace` was initially designed to support arbitrary hierarchy similar to a filesystem, but we limited the initial lookup support to only 2 levels (database, table). I think you might be aware of this.
</blockquote></details>

- about ranges: currently every descriptor ID, be it db or table, causes a range to be allocated.
- about leases: the fact there are no leases on dbs currently is a bug and we are lucky our users do not notice because traffic on db descriptors is low.

  In contrast once schemas are available there will be many more create/drop traffic on schemas, also inside transactions, and users will be much more sensitive to the lack of proper transactional semantics on those. I don't want to be casual with them as we've been with databases.

- about system.namespace and hierarchy. We're making a SQL database with at most 2-3 levels. The full generality of system.namespace was premature design IMHO. Also even if we wanted to implement arbitrary deep objects (say, if/when we want a filesystem) this schema is poor because it requires recursive queries to normalize/resolve paths. There are better approaches to storing filesystem paths, for example that used in ReiserFS or ZFS, and if/when we want to do that we should get inspiration from there.

Until then I would like to avoid considering this theoretical situation in this discussion, and instead capitalize/utilize the fact we need at most 2-3 levels to optimize/simplify things.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuISzR12R90pd3ly3Q:bod6epw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
> about ranges: currently every descriptor ID, be it db or table, causes a range to be allocated.

Right, but I don't think the addition of schemas will cause this problem to get much worse. We should fix this for databases.

> In contrast once schemas are available there will be many more create/drop traffic on schemas, also inside transactions, and users will be much more sensitive to the lack of proper transactional semantics on those. I don't want to be casual with them as we've been with databases.

Hmm, ok. Can you elaborate on these concerns a bit more? I'm not up to speed on what problems the lack of database schemas is causing. Also, cc @vivekmenezes who has been thinking about "schema leases" for 2.2. I'm not quite sure what level of "schema" he is thinking of.

> Until then I would like to avoid considering this theoretical situation in this discussion, and instead capitalize/utilize the fact we need at most 2-3 levels to optimize/simplify things.

Ack. I didn't mean to suggest that we should support more levels than database / schema / table.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuKhwMEeODt-G0-mdS:bkcc44u) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
> Right, but I don't think the addition of schemas will cause this problem to get much worse. We should fix this for databases.

"Much" perhaps not, but worse definitely. In test frameworks, schemas get created at a similar rate as tables. So we'd double the traffic on identifiers. This alone actually would make alternative C below more attractive.

> Hmm, ok. Can you elaborate on these concerns a bit more? I'm not up to speed on what problems the lack of database schemas is causing.

I'll look into it, I can't recall the specific problematic SQL patterns out of my head. However perhaps @vivekmenezes knows them by heart?

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuNNp0Bi-Smn1GMmWh:b-x0lasm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
> "Much" perhaps not, but worse definitely. In test frameworks, schemas get created at a similar rate as tables. So we'd double the traffic on identifiers. This alone actually would make alternative C below more attractive.

I'm going to file an issue about this. I wouldn't want these spurious empty ranges to be a reason to prefer one of these alternatives over another. I believe it shouldn't be too difficult to fix. For the purposes of this RFC, I think it should be a minor consideration (though it is great that you pointed it out).

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuSYsY72-cH8_v_vdK:b9xu86h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
> I'm going to file an issue about this.

Note that the issue here is independent of `SchemaDescriptors`. Simply allocating IDs from the same namespace as table IDs and database IDs (as is being suggested in this alternative) will cause the problem.

**Review by vilterp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LNuf8OA6I8UjSa5oJNo:b-sw0cjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> I'm going to file an issue about this.

Note that the issue here is independent of `SchemaDescriptors`. Simply allocating IDs from the same namespace as table IDs and database IDs (as is being suggested in this alternative) will cause the problem. 
</blockquote></details>

issue about databases having empty ranges: https://github.com/cockroachdb/cockroach/issues/24847

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNujoCi44RavjXUSoHb:-LNujoCi44RavjXUSoHc:bcyjnaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L122)):*
> ```Markdown
> 
> - using a simple 1-part name, e.g. `users`, both the "current database"
>   (`SET database` / `USER`) and the *current schema* (`SET
> ```

s/USER/USE/
___
*[docs/RFCS/20181003_schemas.md, line 224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNulLgRDgcK3FyL8Nnc:-LNulLgRDgcK3FyL8Nnd:b-p8zg3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L224)):*
> ```Markdown
>   inherit directly from database to table even with schemas defined.
> 
>   The ability to customize zones per schema would be a feature that
> ```

I think this is worth doing at the same time that we introduce user-defined schemas. 
___
*[docs/RFCS/20181003_schemas.md, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNupHonCajxHPMCKsgp:-LNupHonCajxHPMCKsgq:b3t83r9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L362)):*
> ```Markdown
> latency.
> 
> ### Alternative B: schemas with descriptors
> ```

I prefer option B. 

I dislike option A because it creates such a difference between databases (which have descriptors) and schemas (which have some other to-be-determined place to store the things that are currently in the DB descriptor). Option A would be more appealing to me if we planned to move away from database descriptors too and make them use the same system as for schemas.

Option C is acceptable, but it feels hackier to me than option A or B. I think the concerns about database descriptor leases are more severe for option C (for option B, operations on schemas would not be substantially more common or more complex than operations on databases are today, so it doesn't make things much worse, but in option C the kinds of changes we'd make to database descriptors become more complex). Vivek's schema leases should take care of this no matter which alternative we choose here, though.

**Comment by knz** [comment]
> operations on schemas would not be substantially more common or more complex than operations on databases are today, 

that's not quite corrects -- we've found that test frameworks create/destroy schemas at a similar rate as tables, not databases.

**Review by vivekmenezes — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LOTyh162wfaTlDphA91:-LOTyh17CLIVvji0JoSt:brhg7fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L31)):*
> ```Markdown
> A. compatibility with PostgreSQL -- ORMs and client apps expect this to
>    work.
> B. simplify the management of multi-user or multi-app clusters.
> ```

I'm not convinced that B is important at this stage. Is A a priority?
___
*[docs/RFCS/20181003_schemas.md, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LOU-dIyC9Oz-LyBljpU:-LOU-dIyC9Oz-LyBljpV:bbjfsng) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L356)):*
> ```Markdown
> (for example, privileges and grants for creating new tables) would be
> stored in a separate system table. (to be defined by further experimentation)
> ```

i think privileges should be kept out of descriptors. See https://github.com/cockroachdb/cockroach/issues/2939

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LOTyh162wfaTlDphA91:-LOUvfTwALtyYnzxCLoG:bxtlc5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L31)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I'm not convinced that B is important at this stage. Is A a priority?
</blockquote></details>

Robert has pressed for A, it is found to be a blocker in the sales pipeline

Lakshmi has pressed for B, it is found to be demand from managed users

The motivation for this work (and the fact it was put on the 2.2 roadmap) comes from them. We can ask Andy Woods for clarification.
___
*[docs/RFCS/20181003_schemas.md, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LOU-dIyC9Oz-LyBljpU:-LOUw-H49nMDWcaQM5vZ:b-ud4x2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L356)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

i think privileges should be kept out of descriptors. See https://github.com/cockroachdb/cockroach/issues/2939
</blockquote></details>

I tend to agree -- although that seems to be an orthogonal concern to this RFC. As long as table/db descs are storing descriptors, any new schema descriptors can do the same; only when we actively work to remove privs from db/table descs, should we remove them from schema descs too. The design for the two things ought to be consistent.

I will reword this to express this more clearly.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LOVr7rf2u2a6s4SqUyy:b-topfie) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

issue about databases having empty ranges: https://github.com/cockroachdb/cockroach/issues/24847
</blockquote></details>

FYI, databases and views will no longer create empty ranges. So the "allocation of ranges" part here is now FUD. :smiley:

**Review by vilterp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LOVt4Uo9BUDIqMAhMPd:b-8u1s55) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, databases and views will no longer create empty ranges. So the "allocation of ranges" part here is now FUD. :smiley: 
</blockquote></details>

Saw that. Nice work @ridwanmsharif!

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20181003_schemas.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNujoCi44RavjXUSoHb:-LVJFw46-NU8syTbHvCo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L122)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/USER/USE/
</blockquote></details>

Done.
___
*[docs/RFCS/20181003_schemas.md, line 224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNulLgRDgcK3FyL8Nnc:-LVJG3ds-N9AdSRfJ4h4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L224)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this is worth doing at the same time that we introduce user-defined schemas. 
</blockquote></details>

Done.
___
*[docs/RFCS/20181003_schemas.md, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNu6Vcg7NwJj6fp9z4x:-LVJG6jXCOkRIgzPuw5C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L280)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Saw that. Nice work @ridwanmsharif!
</blockquote></details>

Done.
___
*[docs/RFCS/20181003_schemas.md, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LOU-dIyC9Oz-LyBljpU:-LVJGt_wC-Ny_qkwn4NU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L356)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I tend to agree -- although that seems to be an orthogonal concern to this RFC. As long as table/db descs are storing descriptors, any new schema descriptors can do the same; only when we actively work to remove privs from db/table descs, should we remove them from schema descs too. The design for the two things ought to be consistent.

I will reword this to express this more clearly.
</blockquote></details>

Done.
___
*[docs/RFCS/20181003_schemas.md, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-LNupHonCajxHPMCKsgp:-LVJH5uz2drqcy4gPsVQ:b-n7i01j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8628ec78b3e0b10e9c5fde9a286e31a4f8995df2/docs/RFCS/20181003_schemas.md#L362)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I prefer option B. 

I dislike option A because it creates such a difference between databases (which have descriptors) and schemas (which have some other to-be-determined place to store the things that are currently in the DB descriptor). Option A would be more appealing to me if we planned to move away from database descriptors too and make them use the same system as for schemas.

Option C is acceptable, but it feels hackier to me than option A or B. I think the concerns about database descriptor leases are more severe for option C (for option B, operations on schemas would not be substantially more common or more complex than operations on databases are today, so it doesn't make things much worse, but in option C the kinds of changes we'd make to database descriptors become more complex). Vivek's schema leases should take care of this no matter which alternative we choose here, though. 
</blockquote></details>

I have addded your reaction to the rationale section below.

**Review by knz — COMMENTED** [review]
Discussed/studied this with @bobvauwter today. Came up with new alternative D; also found pitfalls due to mixed-version clusters with the other alternatives and the new one. Came up with a test scenario to implement as foundation for the evaluation of the various alternatives.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30916#-:-LVJV9JY6tmpUpc4TSLw:bc27ua5)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by Disturbing** [comment]
Waiting for this one to go live. Any ETA?

**Comment by knz** [comment]
No earlier than 19.2
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by jordanlewis** [comment]
@Disturbing, out of curiosity, what's your use case? In CockroachDB, there's little difference between a `CREATE DATABASE` statement and what a `CREATE SCHEMA` would eventually do. Could you make do with a `DATABASE`?

**Comment by Disturbing** [comment]
Im working on using hasura's graphql engine @ global scale.  Last mentioned, there's an error here: https://github.com/hasura/graphql-engine/issues/411#issuecomment-444380438

I can ask their team to change the setup if you have a recommendation.

I personally haven't ran the startup just yet, but I should be able to find some time to repo this error if it still exists or not.

**Comment by jordanlewis** [comment]
@Disturbing my recommendation would be to use databases instead of schemas and see whether that satisfies your use case.

**Comment by knz** [comment]
Updated RFC with new version numbers, also minor simplification informed by #41482

**Comment by otan** [comment]
hey @knz is this still up to date given the temp table work? i'm not sure what the follow up is given that schemas have an `id` in the system.namespace table, but there is no metadata work. are we looking to do something along the lines of `B`?

furthermore, there is a performance issue we need to solve with table resolution given https://github.com/cockroachdb/cockroach/issues/44802

**Comment by ajwerner** [comment]
I like B best, but as you note, we'll need a story around caching and retrieval. I would like to move databases and schemas to be leased resources like tables. There are lots of details to be worked out but in my mind:

A database version will change when adding, deleting, or renaming a child. Same for a schema. 

We'll also need a migration to move the tables to a public schema. 

There's lots of details to work out here. This is all part of planning work for the upcoming sql schema team.

**Comment by knz** [comment]
This draft RFC has not been updated since Arul chose his scheme for the new system.namespace. Now that this choice is made, some of the discussion in the RFC can be collapsed.


is this a discussion we want to pick back up now?

**Comment by rohany** [comment]
It looks like user defined schemas will be an item for the sql features team in 20.2, so its time for us to start thinking about this again.

**Comment by ajwerner** [comment]
> It looks like user defined schemas will be an item for the sql features team in 20.2, so its time for us to start thinking about this again.

Yep, also on the schema teams plate. The tentative breakdown is that name resolution and the associated leasing falls on the schema team. Let's make sure to coordinate.

**Comment by lucy-zhang** [comment]
For anyone following this issue, there's an updated RFC for user-defined schemas here: https://github.com/cockroachdb/cockroach/pull/48276

**Comment by lucy-zhang** [comment]
At knz's suggestion I am marking this version of the RFC as "superseded" and merging it.

bors r+
